src/core/lib/iomgr/event_engine/uv/impl.cc

```diff
@@ -0,0 +1,1301 @@
+absl::Status LibuvEndpoint::Connect(
+    LibuvEventEngine* engine,
+    grpc_event_engine::experimental::EventEngine::OnConnectCallback on_connect,
+    const grpc_event_engine::experimental::EventEngine::ResolvedAddress& addr) {
+  on_connect_ = std::move(on_connect);
+  engine->Schedule([addr, this](LibuvEventEngine* engine) {
+    uvTCP_->RegisterUnsafe(engine);
+    int r = uv_tcp_connect(
+        &connect_, &uvTCP_->tcp_, addr.address(),
+        [](uv_connect_t* req, int status) {
+          LibuvEndpoint* epRaw = reinterpret_cast<LibuvEndpoint*>(req->data);
+          std::unique_ptr<LibuvEndpoint> ep(epRaw);
+          auto on_connect = std::move(ep->on_connect_);
+          if (status == 0) {
+            ep->PopulateAddressesUnsafe();
+            if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+              gpr_log(GPR_DEBUG, "LibuvEndpoint@%p::Connect, success",
+                      epRaw->uvTCP_);
+            }
+            on_connect(std::move(ep));
+          } else {
+            if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+              gpr_log(GPR_INFO, "LibuvEndpoint@%p::Connect, failed: %i",
+                      epRaw->uvTCP_, status);
+            }
+            on_connect(absl::UnknownError(
+                "uv_tcp_connect gave us an asynchronous error"));
+          }
+        });
+    // If we fail the call to uv_connect, the handle won't even be monitored by
+    // the libuv, and won't need to be closed (it won't be in an opened state).
+    // We can bail early here.
+    if (r != 0) {
+      auto on_connect = std::move(on_connect_);
+      if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+        gpr_log(GPR_INFO, "LibuvEndpoint@%p::Connect, failed: %i", uvTCP_, r);
+      }
+      delete this;
+      on_connect(absl::UnknownError("uv_tcp_connect gave us an error"));
+    }
+  });
+  return absl::OkStatus();
+}
+
+void LibuvWrapperBase::RegisterUnsafe(LibuvEventEngine* engine) {
+  auto success = uv_tcp_init(engine->GetLoop(), &tcp_);
+  GPR_ASSERT(success == 0);
+}
+
+LibuvEndpointWrapper::LibuvEndpointWrapper(
+    const grpc_event_engine::experimental::EndpointConfig& args,
+    std::unique_ptr<grpc_event_engine::experimental::SliceAllocator>
+        slice_allocator)
+    : args_(args), slice_allocator_(std::move(slice_allocator)) {
+  write_req_.data = this;
+}
+
+LibuvListener::LibuvListener(
+    Listener::AcceptCallback on_accept,
+    grpc_event_engine::experimental::EventEngine::Callback on_shutdown,
+    const grpc_event_engine::experimental::EndpointConfig& args,
+    std::unique_ptr<grpc_event_engine::experimental::SliceAllocatorFactory>
+        slice_allocator_factory)
+    : uvTCP_(new LibuvListenerWrapper(on_accept, on_shutdown, args,
+                                      std::move(slice_allocator_factory))) {
+  uvTCP_->tcp_.data = uvTCP_;
+}
+
+LibuvListener::~LibuvListener() { uvTCP_->Close(GetEventEngine()); }
+
+absl::StatusOr<int> LibuvListener::Bind(
```

Comment:
Just leaving this comment here to seek some clarification. As per my understanding, the Bind method can be called multiple times with different addresses but it seems like there is only one uv_tcp_t handle per LibuvListener here. Not sure if libuv internally creates and maintains multiple sockets per uv_tcp_t handle or if there should be a linked list of uv_tcp_t handles for each LibuvListener